### PR TITLE
Edit a contact route

### LINKED ISSUE #25: Edit a contact
Edit a contact with a PUT request

Flask documentation on [Flask Blueprints](https://flask.palletsprojects.com/en/1.1.x/blueprints/#blueprints)
An example of the REST endpoints - [https://github.com/celinehuang/kom_get_your_bucha/blob/master/backend/app/items/views.py]()

### PR DESCRIPTION
**Pull request by dgiltz**
# Summary

Edit a contact route code
	
## Test Plan

Tests will be merged in second pull request

## Related Issues

Which issues does this PR resolve/work on?
Closes #25

### CODE REVIEW

File: backend/app/contact/routes.py [hunk 1]
```diff
@@ -14,3 +14,55 @@ def delete_contact_type(id):
     db.session.commit()
 
     return jsonify(contact.serialize)
+
+@contact.route("/<uuid:id>/edit", methods=["PUT"])
```
Comment (CurtisMIT) [inline backend/app/contact/routes.py:18]: @dgiltz Was wondering whether you could change the route name to `"/<uuid:id>"`? Feel free to checkout our discussion on #99.

File: backend/app/contact/routes.py [hunk 2]
```diff
@@ -14,3 +14,55 @@ def delete_contact_type(id):
...
+        contact.neighbourhood = neighbourhood
+
+    if projects is not None:
+        contact.projects = projects
+
+    if contact is not None:
+        contact.contact_type = contact.contact_type
+
+    if not data:
+        abort(400, "No fields to update")
```
Comment (CurtisMIT) [inline backend/app/contact/routes.py:63]: @dgiltz We had a discussion regarding whether to check for `{"name" : ""}` and `{}` and ended up deciding to check for both. Was wondering whether you could add the check for `{"name": ""}` as well? 
An example:
https://github.com/hack4impact-mcgill/mu-crm-tool/blob/main/backend/app/project/routes.py#L21

File: backend/tests/test_contact.py [hunk 3]
```diff
@@ -56,3 +56,47 @@ def test_delete_a_contact(self):
...
+        c = Contact(
+            id=c_id,
+            name="dummy name",
+            email="dummy email",
+            secondary_email="dummy secondary email",
+            cellphone="dummy cellphone",
+            role="dummy role",
+            organization="dummy organization",
+            neighbourhood="dummy neighbourhood",
+            contact_type=dummy_ct,
```
Comment (CurtisMIT) [inline backend/tests/test_contact.py:79]: @dgiltz The reason why your tests are failing, is because at this line you are assigning a `ContactType` rather than the `uuid` of the `ContactType`. However, I think even if you were to assign the `uuid` of the `ContactType`, it would still fail. The reason is you probably want to assign `dummy_ct` to `c.contact_types` instead, this is due to the `backref` defined within the `ContactType` model.

Links:
https://github.com/hack4impact-mcgill/mu-crm-tool/blob/main/backend/app/models.py#L159
https://flask-sqlalchemy.palletsprojects.com/en/2.x/models/#one-to-many-relationships

File: backend/tests/test_contact.py [hunk 4]
```diff
@@ -56,3 +56,47 @@ def test_delete_a_contact(self):
...
+        self.assertDictContainsSubset(update_obj, json_response)
+
+        # edit a contact that does not exist
+        response = self.client.put(
+            "/contact/{}/edit".format(uuid.uuid4()),
+            content_type="application/json",
+            data=json.dumps({}),
+        )
+        self.assertEqual(response.status_code, 404)
+
```
Comment (CurtisMIT) [inline backend/tests/test_contact.py:102]: @dgiltz if possible, can you add a test for an edit request that has zero valued parameters? Like `{"name": ""}`.

https://github.com/hack4impact-mcgill/mu-crm-tool/blob/main/backend/tests/test_project.py#L283

**Review by CurtisMIT — CHANGES_REQUESTED** [review]
@dgiltz I think you forgot to format your code with `black`

File: backend/app/contact/routes.py [hunk 5]
```diff
@@ -1,8 +1,45 @@
...
+        secondary_email=secondary_email,
+        cellphone=cellphone,
+        role=role,
+        organization=organization,
+        neighbourhood=neighbourhood,
+        projects=projects,
+    )
+
+    db.session.add(new_contact)
+    db.session.commit()
```
Comment (CurtisMIT) [inline backend/app/contact/routes.py:40]: @dgiltz why is this here? There is no need for this to be here as you have a separate pr for the create a contact route.

File: backend/app/contact/routes.py [hunk 6]
```diff
@@ -14,3 +51,71 @@ def delete_contact_type(id):
...
+    if organization is not None and organization != "":
+        contact.organization = organization
+
+    if neighbourhood is not None and neighbourhood != "":
+        contact.neighbourhood = neighbourhood
+
+    if contact_type is not None and contact_type != "":
+        contact.contact_type = contact.contact_type
+
+    if not data:
```
Comment (CurtisMIT) [inline backend/app/contact/routes.py:115]: @dgiltz this is redundant I think as you've already checked for this condition above.

**Review by CurtisMIT — CHANGES_REQUESTED** [review]
@dgiltz is it possible for you to rebase on top of the most update master? It's kinda hard to review if there's all these changes that are not relevant to your pr. Thanks!

File: backend/app/contact/routes.py (hunk 5 above)
Comment (dgiltz) [reply backend/app/contact/routes.py:40]: Sorry, problems with merging the code and my editor caused a lot of extraneous stuff to slip through

**Comment by dgiltz** [comment]
Obselete
